### PR TITLE
feat: replicate support

### LINKED ISSUE #242: 🚀 Feature: Replicate Instrumentation
### Which component is this feature for?

All Packages

### 🔖 Feature description

Instrument calls to [Replicate](https://replicate.com/docs/get-started/python), including adding attributes for input parameters, model, etc. - similarly to our [Anthropic](https://github.com/traceloop/openllmetry/tree/main/packages/opentelemetry-instrumentation-anthropic) instrumentation. The instrumentation should support all types of calls - streaming, non streaming, async, etc.

### 🎤 Why is this feature needed ?

Completeness of OpenLLMetry

### ✌️ How do you aim to achieve this?

Similar to other instrumentations we have in this repo.

### 🔄️ Additional Information

* Make sure to initialize this in [Traceloop SDK](https://github.com/traceloop/openllmetry/blob/main/packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py#L241).
* Add proper [tests](https://github.com/traceloop/openllmetry/tree/main/packages/traceloop-sdk/tests).

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Kartik1397**
closes #242 

/claim #242

### CODE REVIEW

File: packages/opentelemetry-instrumentation-replicate/opentelemetry/instrumentation/replicate/__init__.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+
+def _set_span_attribute(span, name, value):
+    if value is not None:
+        if value != "":
+            span.set_attribute(name, value)
+    return
+
+
+def _set_input_attributes(span, kwargs):
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-replicate/opentelemetry/instrumentation/replicate/__init__.py:56]: This depends on the exact model we use. What I'd want to see here is some "dynamic" way of extracting the arguments and sending them as span attributes, with things like temperature indeed in the semantic convention names if they exist. This is an example for Llama:
https://replicate.com/meta/llama-2-70b-chat/api?tab=python

File: packages/opentelemetry-instrumentation-replicate/opentelemetry/instrumentation/replicate/__init__.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    _set_span_attribute(span, SpanAttributes.LLM_REQUEST_MODEL, kwargs.get("model"))
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-replicate/opentelemetry/instrumentation/replicate/__init__.py:57]: This is incorrect. Model will always be empty

**Comment by nirga** [comment]
Thanks so much for this @Kartik1397! Wrote a couple of comments. I'd also suggest actually testing this against app.traceloop.com because I'm not sure it will work as expected currently. You can build a new simple sample app for that.

**Review by nirga — CHANGES_REQUESTED** [review]
Thanks so much for this @Kartik1397! Wrote a couple of comments. I'd also suggest actually testing this against app.traceloop.com because I'm not sure it will work as expected currently. You can build a new simple sample app for that.

**Comment by Kartik1397** [comment]
Thanks, @nirga for reviewing the PR! I've incorporated your feedback in my latest commit. Please let me know if there's still anything missing or can be done in better way. This is my first time using OpenTelemetry SDK, I may have overlooked some straightforward aspects. Your guidance is greatly appreciated!

Testing on api.traceloop.com

https://github.com/traceloop/openllmetry/assets/24036721/1df68b94-5fb8-4613-a38a-21d08f82b566

**Comment by nirga** [comment]
Perfect! I see that `poetry install` is failing because of a problem with the lock file (see checks below), other than that if the CI works afterwards we're good to go.
Thanks! 🙏

File: packages/traceloop-sdk/tests/test_replicate_instrumentaion.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def test_replicate_image_generation(exporter):
+    @task(name="image_generation")
+    def generate_image():
+        image = replicate.run(
+            "stability-ai/stable-diffusion:27b93a2413e7f36cd83da926f3656280b2931564ff050bf9575f1fdf9bcd7478",
+            input={"prompt": "robots"}
+        )
+        print(image)
```
Comment (nirga) [inline packages/traceloop-sdk/tests/test_replicate_instrumentaion.py:12]: nit: remove

File: packages/opentelemetry-instrumentation-replicate/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+description = "Automatically generated by Nx."
+authors = [ ]
+license = 'Proprietary'
+readme = 'README.md'
+
+  [[tool.poetry.packages]]
+  include = "opentelemetry/instrumentation/replicate"
+
+  [tool.poetry.dependencies]
+  python = ">=3.9,<3.11"
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-replicate/pyproject.toml:24]:
```suggestion
  python = ">=3.9,<4"
```

**Review by nirga — COMMENTED** [review]
Perfect! Can you just add the sample app you used?

File: packages/traceloop-sdk/tests/test_replicate_instrumentaion.py (hunk 3 above)
Comment (Kartik1397) [reply packages/traceloop-sdk/tests/test_replicate_instrumentaion.py:12]: @nirga Should I remove all print statement from this test file?

**Comment by Kartik1397** [comment]
> Perfect! Can you just add the sample app you used?

Should I add it to https://github.com/traceloop/openllmetry/tree/main/packages/sample-app/sample_app?

**Comment by nirga** [comment]
> > Perfect! Can you just add the sample app you used?
> 
> 
> 
> Should I add it to https://github.com/traceloop/openllmetry/tree/main/packages/sample-app/sample_app?

Yes!

**Comment by Kartik1397** [comment]
Done @nirga. Also reduced minimum python version required for  `opentelemetry-instrumentation-replicate` to 3.8 as replicate support python 3.8. https://pypi.org/project/replicate/

File: packages/traceloop-sdk/pyproject.toml [hunk 5]
```diff
@@ -45,12 +45,14 @@ opentelemetry-instrumentation-chromadb = {path="../opentelemetry-instrumentation
...
 opentelemetry-instrumentation-llamaindex = {path="../opentelemetry-instrumentation-llamaindex", develop=true}
 opentelemetry-instrumentation-bedrock = {path="../opentelemetry-instrumentation-bedrock", develop=true}
+opentelemetry-instrumentation-replicate = {path="../opentelemetry-instrumentation-replicate", develop=true}
 colorama = "^0.4.6"
 tenacity = "^8.2.3"
 pydantic = "^2.5.0"
 jinja2 = "^3.1.2"
 deprecated = "^1.2.14"
 posthog = "^3.0.2"
+replicate = "^0.22.0"
```
Comment (nirga) [inline packages/traceloop-sdk/pyproject.toml:55]: This shouldn't be a dependency of the SDK, just a test dependency

File: packages/traceloop-sdk/tests/test_replicate_instrumentaion.py (hunk 3 above)
Comment (nirga) [reply packages/traceloop-sdk/tests/test_replicate_instrumentaion.py:12]: @Kartik1397 yes, these aren't printed anyway

**Comment by nirga** [comment]
@Kartik1397 when installing the SDK there's a warning
`Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.` (see checks logs)

**Comment by Kartik1397** [comment]
> @Kartik1397 when installing the SDK there's a warning
Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run poetry lock [--no-update] to fix it. (see checks logs).

This should be fixed now.

> Also, I'm not sure your code will work for streaming - have you tested it?

I can see traces of stream function calls are getting consumed by traceloop(with propmpts). Do we need to capture response in case of streaming API?

Screenshot 2023-12-18 at 7 54 24 PM

```
{
    "name": "replicate.stream",
    "context": {
        "trace_id": "0x0e03350bb8f375674076a774eb4d964f",
        "span_id": "0x84224c4e5915e5b4",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": "0x21c7475439b11a24",
    "start_time": "2023-12-18T14:36:54.448065Z",
    "end_time": "2023-12-18T14:36:54.448092Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "llm.vendor": "Replicate",
        "llm.request.type": "completion",
        "traceloop.workflow.name": "chat_stream_generator",
        "llm.request.model": "meta/llama-2-70b-chat:02e509c789964a7ea8736978a43525956ef40397be9033abf9fd2badfe68c9e3",
        "llm.prompts.0.user": "tiny robot"
    },
    "events": [],
    "links": [],
    "resource": {
        "attributes": {
            "service.name": "sample_app/replicate_streaming.py"
        },
        "schema_url": ""
    }
}
```

**Comment by nirga** [comment]
> > @Kartik1397 when installing the SDK there's a warning
> > Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run poetry lock [--no-update] to fix it. (see checks logs).
> 
> This should be fixed now.
> 
> > Also, I'm not sure your code will work for streaming - have you tested it?
> 
> I can see traces of stream function calls are getting consumed by traceloop(with propmpts). Do we need to capture response in case of streaming API?
> 
> Screenshot 2023-12-18 at 7 54 24 PM
> ```
> {
>     "name": "replicate.stream",
>     "context": {
>         "trace_id": "0x0e03350bb8f375674076a774eb4d964f",
>         "span_id": "0x84224c4e5915e5b4",
>         "trace_state": "[]"
>     },
>     "kind": "SpanKind.CLIENT",
>     "parent_id": "0x21c7475439b11a24",
>     "start_time": "2023-12-18T14:36:54.448065Z",
>     "end_time": "2023-12-18T14:36:54.448092Z",
>     "status": {
>         "status_code": "OK"
>     },
>     "attributes": {
>         "llm.vendor": "Replicate",
>         "llm.request.type": "completion",
>         "traceloop.workflow.name": "chat_stream_generator",
>         "llm.request.model": "meta/llama-2-70b-chat:02e509c789964a7ea8736978a43525956ef40397be9033abf9fd2badfe68c9e3",
>         "llm.prompts.0.user": "tiny robot"
>     },
>     "events": [],
>     "links": [],
>     "resource": {
>         "attributes": {
>             "service.name": "sample_app/replicate_streaming.py"
>         },
>         "schema_url": ""
>     }
> }
> ```

Yes, of course :) Same as we do in OpenAI.

**Comment by Kartik1397** [comment]
Okay, will add support for streaming.

**Comment by Kartik1397** [comment]
Pushed changes to capture response streams.

**Review by nirga — APPROVED** [review]

**Comment by nirga** [comment]
Failed as Github doesn't recognize newly added Replicate API token. Tested locally.
Merging.

**Pull request merged**